Shipping labels M3 tracks

0nko opened the pull request:
This PR adds event tracking to shipping label creation flow. The following events are tracked:

- The main flow steps completed (action events)
- Address validation results (action events)
- Button taps (view events)
- Spinner taps (view events)

**To test:**
1. Select an order eligible for SL creation
2. Go through the purchase flow
3. Verify each step is tracked (**shipping_label_purchase_flow**: `state` = `[started, origin_address_complete, destination_address_complete, packages_selected, carrier_rates_selected, customs_complete, purchase_initiated, purchase_succeeded, purchase_failed]`
4. Verify the purchase `amount` is tracked with the `purchase_initiated` step

~Note: This PR is a draft until the purchase functionality is implemented.~

0nko committed:
Add the create shipping label flow tracking events

0nko committed:
Add the purchase shipping label button click listener

0nko committed:
Add flow step tracking

0nko committed:
Add address editing tracking

0nko committed:
Add shipping carriers view events

0nko committed:
Add address suggestions view events

0nko committed:
Track packages view events

0nko committed:
Add payment method view events

0nko committed:
Add address validation result tracking

hichamboushaba commented:
@0nko the PR #3765 for purchase functionality has been merged, so you can go forward with your implementation here 👍

0nko commented:
@hichamboushaba Got it, thanks 👍

0nko committed:
Merge branch 'develop' into issue/sl-flow-event-tracking

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelsStateMachine.kt

0nko committed:
Add tracking for starting the purchase

0nko committed:
Add purchase amount, success and failure event tracking

0nko committed:
Fix ktlint issues

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt:
```diff
@@ -198,12 +200,42 @@ class CreateShippingLabelViewModel @AssistedInject constructor(
+    )
+
+    private fun trackPurchaseInitiated(amount: BigDecimal) {
+        AnalyticsTracker.track(
+            Stat.SHIPPING_LABEL_PURCHASE_FLOW,
+            mapOf(
+                "state" to "purchase_initiated",
+                "amount" to amount.toPlainString()
+            )
+        )
```
I'm wondering if it won't be useful to track the usage of the checkbox to fulfill the order automatically here?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt:
```diff
@@ -378,6 +413,7 @@ class CreateShippingLabelViewModel @AssistedInject constructor(
                     triggerEvent(ShowSnackbar(R.string.shipping_label_create_purchase_fulfill_error))
                 }
             }
+            AnalyticsTracker.track(Stat.SHIPPING_LABEL_PURCHASE_FLOW, mapOf("state" to "purchase_succeeded"))
```
I remember we discussed regarding the async nature of the purchase, and whether we should change the UI to align with it or not, WDYT about tracking the duration of the purchase API call here, and add it as a property to the event?

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/CreateShippingLabelViewModel.kt:
```diff
@@ -378,6 +413,7 @@ class CreateShippingLabelViewModel @AssistedInject constructor(
                     triggerEvent(ShowSnackbar(R.string.shipping_label_create_purchase_fulfill_error))
```
I think we should add an event to track the failure of fulfilling too, WDYT?

hichamboushaba reviewed (commented):
Nice work, I tried testing all the events, and they work as expected.
I left some comments suggesting some small changes.
And I wanted to ask about the process to [register events](P91TBi-3Py-p2), should it be done before releasing the feature or later?

edit: one last point that I forgot to mention, in the PR description the step is said to be part of the property `action`, but it's in `state`, I checked in the spreadsheet, and it correctly says `state`, so I assume it's just a typo in the description, or it was changed after creating the PR.

0nko committed:
Merge branch 'develop' into issue/sl-flow-event-tracking

0nko committed:
Track the fullfill-order option

0nko committed:
Use string constants instead of literals in tracking

0nko committed:
Track order fulfill result

0nko committed:
Track SL purchase request duration

0nko committed:
Fix ktlint issue

0nko commented:
>  one last point that I forgot to mention, in the PR description the step is said to be part of the property action, but it's in state

It used to be `action` but `state` made more sense for some "states" :) 

Thanks for the review @hichamboushaba. I've addressed all comments and the PR's ready for another round.

hichamboushaba commented:
Thanks @0nko for taking the time to address my comments, they work as expected.
But I found an issue that I missed in my initial review, when testing the app, I noticed that the order of events `purchase_initiated` and `purchase_succeeded` is not correct:
```
2021-04-12 15:43:22.951 12615-12615/com.woocommerce.android.dev I/WooCommerce-UTILS: 🔵 Tracked: shipping_label_purchase_flow, Properties: {"state":"purchase_succeeded","total_duration":7.419,"blog_id":181362242,"is_wpcom_store":true}
2021-04-12 15:43:22.985 12615-12615/com.woocommerce.android.dev I/WooCommerce-UTILS: 🔵 Tracked: shipping_label_purchase_flow, Properties: {"state":"purchase_initiated","amount":"10.6","fulfill_order":false,"blog_id":181362242,"is_wpcom_store":true}
```
The cause is that `purchaseLabels` is a suspendable function, so the side effect won't be handled until it's completed.
I suggest you move the `purchase_initiated` tracking to the first line of `purcahseLabels` without relying on side effects, WDYT?

0nko committed:
Fix the purchase event tracking order

0nko committed:
Optimize imports

0nko commented:
@hichamboushaba Good catch! It's fixed now.

hichamboushaba reviewed (approved):
Thanks for the fix @0nko, LGTM :shipit:
